horovod/common/gradient_aggregation_v2.py

```diff
@@ -0,0 +1,101 @@
+        # for parameter updates. N is the number of parameters.
+        self.shadow_var = {}
+
+        # Used to know when to allreduce and apply gradients. We allreduce when `self.counter`
+        # is equal to `self.aggregation_frequency`. We apply gradients when `self.counter` is
+        # equal to 0.
+        self.counter = tf.Variable(initial_value=0)
+
+        self._sparse_as_dense = sparse_as_dense
+
+        # A dictionary containing the shape of each grad.
+        # This is used when gradient aggregation frequency > 1 and
+        # there are grads that have dynamically set shapes.
+        self.grad_updated_sizes_dict = grad_updated_sizes_dict
+
+        # Contains the mapping of indexes of grad updates that are
+        # not None to their index in shadow vars which only
+        # contains not None gradients. When performing gradient
+        # aggregation we have to remove them from the list of grads
+        # prior passing them into a tf.cond().
+        self.not_none_indexes = {}
+        self.num_none_grad_updates = 0
+
+    @tf.function
+    def compute_gradients(self, grads):
+        resulting_grads = []
+        for idx, grad in enumerate(grads):
+            if idx not in self.shadow_var.keys():
+                if grad is not None:
+                    self.shadow_var[idx] = tf.Variable(
+                        initial_value=tf.zeros_like(grad),
+                        trainable=False,
+                        dtype=grad.dtype,
+                    )
+                else:
+                    self.num_none_grad_updates += 1
+                    continue
+            if grad is not None:
+                self.shadow_var[idx].assign_add(grad)
+                resulting_grads.append(self.shadow_var[idx].read_value())
+
+        assert len(self.shadow_var) + self.num_none_grad_updates == len(grads)
+
+        self.counter.assign_add(1)
+        if tf.equal(self.counter, self.aggregation_frequency):
+            resulting_grads = self._allreduce_grads(resulting_grads)
+            assert len(resulting_grads) == len(self.shadow_var)
+            resulting_grads = [
+                resulting_grads[idx] if idx in self.shadow_var else None
+                for idx in range(len(resulting_grads) + self.num_none_grad_updates)
+            ]
+            assert (
+                len(resulting_grads)
+                == len(self.shadow_var) + self.num_none_grad_updates
+            )
+            self._clear_vars()
+
+        return resulting_grads
+
+    def _allreduce_grads(self, grads):
+        allreduced_grads = self._allreduce_grads(grads)
+        if self.aggregation_frequency > 1:
+            gradient_divisor = (
+                self.aggregation_frequency if self.average_aggregated_gradients else 1
+            )
+            allreduced_grads = [
+                grad / gradient_divisor if grad is not None else None for grad in grads
```

Comment:
similar to : https://github.com/determined-ai/horovod/pull/22#discussion_r406932719

Reply:
I didn't see anything about best practices in that link about whether we should use python operators vs tensorflow operators. There is an extensive writeup on the topic here: https://pgaleone.eu/tensorflow/tf.function/2019/05/10/dissecting-tf-function-part-3/#lesson-1-not-all-operators-are-created-equal. 

I do agree with your point though that it seems to apply to mostly inputs, but since who knows what TF does under the covers, all things being equal it seems better to use TF operators vs python operators when possible.

Reply:
I mean in their code snippets they don't use tf operators; however, I do see your point -- happy to move it to tf operators